packages/opencensus-core/src/trace/model/tracer-base.ts

```diff
@@ -0,0 +1,237 @@
+        this, name, kind, traceId, parentSpanId, traceState);
+    return fn(noRecordRootSpan);
+  }
+
+  /** Notifies listeners of the span start. */
+  onStartSpan(root: types.Span): void {
+    if (!this.active) return;
+    if (!root) {
+      return this.logger.debug('cannot start trace - no active trace found');
+    }
+    this.notifyStartSpan(root);
+  }
+
+  /** Notifies listeners of the span end. */
+  onEndSpan(root: types.Span): void {
+    if (!this.active) return;
+    if (!root) {
+      this.logger.debug('cannot end trace - no active trace found');
+      return;
+    }
+    this.notifyEndSpan(root);
+  }
+
+  /**
+   * Registers an end span event listener.
+   * @param listener The listener to register.
+   */
+  registerSpanEventListener(listener: types.SpanEventListener) {
+    this.eventListenersLocal.push(listener);
+  }
+
+  /**
+   * Unregisters an end span event listener.
+   * @param listener The listener to unregister.
+   */
+  unregisterSpanEventListener(listener: types.SpanEventListener) {
+    const index = this.eventListenersLocal.indexOf(listener, 0);
+    if (index > -1) {
+      this.eventListeners.splice(index, 1);
+    }
+  }
+
+  private notifyStartSpan(root: types.Span) {
+    this.logger.debug('starting to notify listeners the start of rootspans');
+    if (this.eventListenersLocal && this.eventListenersLocal.length > 0) {
+      for (const listener of this.eventListenersLocal) {
+        listener.onStartSpan(root);
+      }
+    }
+  }
+
+  private notifyEndSpan(root: types.Span) {
+    this.logger.debug('starting to notify listeners the end of rootspans');
+    if (this.eventListenersLocal && this.eventListenersLocal.length > 0) {
+      for (const listener of this.eventListenersLocal) {
+        listener.onEndSpan(root);
+      }
+    }
+  }
+
+  /**
+   * Starts a span.
+   * @param [options] A SpanOptions object to start a child span.
+   */
+  startChildSpan(options?: types.SpanOptions): types.Span {
+    if (!options || !options.childOf) {
+      this.logger.debug(
+          'no current trace found - must start a new root span first');
+      return new NoRecordSpan();
+    }
+    return options.childOf.startChildSpan(options.name, options.kind);
+  }
+
+  /** Determine whether to sample request or not. */
+  private makeSamplingDecision(options: types.TraceOptions, traceId: string):
+      boolean {
+    // If users set a specific sampler in the TraceOptions, use it.
+    if (options && options.samplingRate !== undefined &&
+        options.samplingRate !== null) {
+      return SamplerBuilder.getSampler(options.samplingRate)
+          .shouldSample(traceId);
+    }
+    let propagatedSample = null;
+    // if there is a context propagation, keep the decision
+    if (options && options.spanContext && options.spanContext.options) {
+      propagatedSample =
+          ((options.spanContext.options & this.IS_SAMPLED) !== 0);
+    }
+
+    let sampleDecision = !!propagatedSample;
```

Comment:
Could this logic be simplified to `return !!propagatedSample || this.sampler.shouldSample(traceId)`?